### PR TITLE
Add `tsh request drop` command

### LINKED ISSUE #14491: Add an easy way to drop a resource access request and return to 'normal' access
## What would you like Teleport to do?

As per https://goteleport.com/docs/enterprise/workflow/resource-requests/#step-78-access-the-requested-resource, if you assume a resource access request, your existing permissions are gone:

> While logged in with a resource access request, users will be blocked from access to any other resources. This is necessary because their certicate now contains an elevated role. Their certificate is restricted to only allow access to the resources they were specifically approved for. They will need to `tsh logout` and log back in to resume their regular access.

Having to run `tsh logout` and then go through the login procedure again just to drop an access request is a little cumbersome.

A command like `tsh drop request` or `tsh request switch-back` or something would be useful to avoid these extra steps.

## What problem does this solve?

Makes resource access requests considerably easier to use from a user perspective.

## If a workaround exists, please include it.

None known. AFAIK `eval $(tsh env)` cannot be used either as it will overwrite the existing credentials for the given cluster/proxy.

### PR DESCRIPTION
**Pull request by nklaassen**
This PR adds the `tsh request drop [<request-id>...]` command which allows users to "drop" access requests, removing them from their cert.

This wasn't really necessary before, because access requests were (almost) always additive: you could always assume more access requests, and they (almost) always gave you more access than you normally have.

With the advent of Resource Access Requests, assuming an access request can now restrict your access to a single resource, not allowing you to list or access anything else. Resource Access Requests also are not additive, so once you assume one you cannot assume another. This leads us to the need to drop access requests so that we can go back to our normal level of access, or assume a different request.

The default when passing `tsh request drop` with no extra arguments is to drop all access requests, you can pass explicit request IDs to only drop those ones (e.g. `tsh request drop 4419d844-04d9-449c-8c5e-db7b739fae0b`)

Example workflow:
```
Nics-MacBook-Pro:teletest nklaassen$ tsh status
> Profile URL:        https://one:4080
  Logged in as:       alice
  Cluster:            cluster-one
  Roles:              requester
  Logins:             -teleport-nologin-efe9b3d5-515f-4e84-9217-673325bc3fc5, -teleport-internal-join
  Kubernetes:         disabled
  Valid until:        2022-07-21 23:57:17 -0700 PDT [valid for 10h32m0s]
  Extensions:         permit-port-forwarding, permit-pty

Nics-MacBook-Pro:teletest nklaassen$ tsh request search --kind node --search auth
Name                                 Hostname Labels Resource ID
------------------------------------ -------- ------ ------------------------------------------------------
5b062385-7497-4c0b-8cb3-9d2ec5e3e07c one-auth        /cluster-one/node/5b062385-7497-4c0b-8cb3-9d2ec5e3e07c

To request access to these resources, run
> tsh request create --resource /cluster-one/node/5b062385-7497-4c0b-8cb3-9d2ec5e3e07c \
    --reason <request reason>

Nics-MacBook-Pro:teletest nklaassen$ tsh request create --resource /cluster-one/node/5b062385-7497-4c0b-8cb3-9d2ec5e3e07c \
    --reason "responding to incident 123"
Creating request...
Request ID: 39620756-5abd-4410-a9d2-ae03ef39ab18
Username:   alice
Roles:      access, node-admins
Resources:  ["/cluster-one/node/5b062385-7497-4c0b-8cb3-9d2ec5e3e07c"]
Reason:     "responding to incident 123"
Reviewers:  [none] (suggested)
Status:     PENDING

hint: use 'tsh login --request-id=<request-id>' to login with an approved request

Waiting for request approval...

Approval received, getting updated certificates...

> Profile URL:        https://one:4080
  Logged in as:       alice
  Active requests:    39620756-5abd-4410-a9d2-ae03ef39ab18
  Cluster:            cluster-one
  Roles:              access, node-admins, requester
  Logins:             nklaassen, root, -teleport-internal-join
  Kubernetes:         disabled
  Allowed Resources:  ["/cluster-one/node/5b062385-7497-4c0b-8cb3-9d2ec5e3e07c"]
  Valid until:        2022-07-21 23:57:17 -0700 PDT [valid for 10h31m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty

Nics-MacBook-Pro:teletest nklaassen$ tsh ls
Node Name Address   Labels
--------- --------- ------
one-auth  [::]:4000

Nics-MacBook-Pro:teletest nklaassen$ tsh request create --roles node-admins
Creating request...
Request ID: 79ebd141-865d-49a5-8f5c-444990ac7946
Username:   alice
Roles:      node-admins
Reason:     [none]
Reviewers:  [none] (suggested)
Status:     PENDING

hint: use 'tsh login --request-id=<request-id>' to login with an approved request

Waiting for request approval...

Approval received, getting updated certificates...

> Profile URL:        https://one:4080
  Logged in as:       alice
  Active requests:    39620756-5abd-4410-a9d2-ae03ef39ab18, 79ebd141-865d-49a5-8f5c-444990ac7946
  Cluster:            cluster-one
  Roles:              access, node-admins, requester
  Logins:             nklaassen, root, -teleport-internal-join
  Kubernetes:         disabled
  Allowed Resources:  ["/cluster-one/node/5b062385-7497-4c0b-8cb3-9d2ec5e3e07c"]
  Valid until:        2022-07-21 23:57:17 -0700 PDT [valid for 10h30m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty

Nics-MacBook-Pro:teletest nklaassen$ tsh ls
Node Name Address   Labels
--------- --------- ------
one-auth  [::]:4000

Nics-MacBook-Pro:teletest nklaassen$ tsh request drop 39620756-5abd-4410-a9d2-ae03ef39ab18
Dropping access request(s): 39620756-5abd-4410-a9d2-ae03ef39ab18...

> Profile URL:        https://one:4080
  Logged in as:       alice
  Active requests:    79ebd141-865d-49a5-8f5c-444990ac7946
  Cluster:            cluster-one
  Roles:              node-admins, requester
  Logins:             root, -teleport-internal-join
  Kubernetes:         disabled
  Valid until:        2022-07-21 23:57:17 -0700 PDT [valid for 10h30m0s]
  Extensions:         permit-port-forwarding, permit-pty

Nics-MacBook-Pro:teletest nklaassen$ tsh ls
Node Name    Address        Labels
------------ -------------- -------------------------
one-auth     [::]:4000
one-node     127.0.0.1:3022 owner=db-admins,test=test
one-node-iot ⟵ Tunnel       owner=db-admins,test=test

```

Fixes #14491

### CODE REVIEW

File: lib/auth/auth_with_roles.go [hunk 1]
```diff
@@ -2077,57 +2076,143 @@ func (a *ServerWithRoles) NewKeepAliver(ctx context.Context) (types.KeepAliver,
...
+	// are intended to reduce allowed permissions so we'll accept them
+	// as-is for now (and ensure the user is allowed to assume them
+	// later).
+	// Note: traits are not currently set for role impersonation.
+	return &services.AccessInfo{
+		Roles: req.RoleRequests,
+	}, nil
+}
+
+// desiredRolesAndTraitsForUser returns the desired roles and traits for a user
```
Comment (strideynet) [inline lib/auth/auth_with_roles.go:2138]: Looks like in refactoring, the function name referenced in this comment didn't get updated.

**Review by strideynet — COMMENTED** [review]
We'll need to make sure we update the docs as well, but happy to see that as a separate PR as this one is already quite large.

**Comment by espadolini** [comment]
Couldn't this be done by having `tsh` store the original login credentials in some other place instead? With that, "dropping" an access request could be done by just restoring the credentials and reissuing with whatever access requests are left (if any).

File: tool/tsh/tsh_test.go [hunk 2]
```diff
@@ -810,21 +810,12 @@ func TestSSHAccessRequest(t *testing.T) {
 	}, setHomePath(tmpHomePath))
 	require.NoError(t, err)
 
-	// log out and back in with no access request
+	// drop the access request
```
Comment (marcoandredinis) [inline tool/tsh/tsh_test.go:813]: What do you think about having the old and the new test?
It seems they target different flows

**Comment by nklaassen** [comment]
> Couldn't this be done by having `tsh` store the original login credentials in some other place instead? With that, "dropping" an access request could be done by just restoring the credentials and reissuing with whatever access requests are left (if any).

@espadolini It could, and I think that idea has a lot of merit. I imagine it would be best to implement it as a "stack" of profiles (per cluster) in `~/.tsh`, change the new command to `tsh request pop`, and have assuming/dropping a request push/pop from the stack.

The downsides I see with that approach are:
1. `tsh` profile management code is looking somewhat complex, I'm not sure I want to fundamentally rearchitect it for this, especially since this will need to backport to v10.
2. Not sure if we should change `~/.tsh` disk layout within a major/minor/patch version.
3. Need to consider all `tsh login`and relogin scenarios - should they just replace the current profile, push to the stack, clear the stack? There's going to be many edge cases.
4. When popping off the stack, need to consider if the old cert is valid/invalid/not there. More to handle here.
5. Will `Connect` need to re-implement this one day, or do they just use `~/.tsh` as well? I don't know.

The current approach just looks like a new relogin with an access request, it does not change anything architecturally, and it's only ~250 changed lines (ignoring protobuf and tests). For simplicity's sake, and since I already wrote it, I'm leaning toward sticking with what I have here, what do you think?

**Comment by nklaassen** [comment]
> We'll need to make sure we update the docs as well, but happy to see that as a separate PR as this one is already quite large.

@strideynet I'll make sure docs get updated, Paul is in the middle of reorganizing access request docs right now so I'll coordinate with him.

File: tool/tsh/tsh_test.go (hunk 2 above)
Comment (nklaassen) [reply tool/tsh/tsh_test.go]: updated the test to leave the old logout flow

File: lib/auth/auth_with_roles.go (hunk 1 above)
Comment (nklaassen) [reply lib/auth/auth_with_roles.go]: thanks, fixed

**Review by marcoandredinis — APPROVED** [review]
LGTM

One question though
Is it a requirement to lose all other access when receiving a Resource Access?

**Review by strideynet — APPROVED** [review]

**Comment by nklaassen** [comment]
> One question though Is it a requirement to lose all other access when receiving a Resource Access?

@marcoandredinis In the current implementation, it is. Historically your certificate always contains a list of roles which the user had at the time the cert was issued, and these would always be used to evaluate access. When assuming a role access requests, this role would be added to your cert, and you could carry on as if you had that role. When designing Resource Access Requests, we decided to augment this by adding a list of resource IDs which the cert would be able to access, to exclude access to all other resources, so that the elevated role could not be used to access anything else.

You could imagine an alternative scheme where we encode in the cert some mapping of which extra roles you effectively "have" when evaluating access to a given resource, in addition to your base roles. This could make Resource Access Requests additive, and probably would improve UX, but would be somewhat more complicated and could be more likely to make certs grow too large. Might be worth doing in the future though

**Comment by marcoandredinis** [comment]
@nklaassen Thank you for the explanation.
I was trying to come up with a solution for this use case, and my first (hypothetical) approach would be to have some kind of Just In Time Roles.
Roles that would be created in the context of a resource access request with access to the required resource(s).
Only the requesting user could assume that role and they would be destroyed afterwards.

Not sure how that would play out with the rest of the system :thinking: 
Well, a bit late to consider this approach (maybe it was considered), and I'm sure it would have a different set of issues

**Comment by nklaassen** [comment]
> @nklaassen Thank you for the explanation. I was trying to come up with a solution for this use case, and my first (hypothetical) approach would be to have some kind of Just In Time Roles. Roles that would be created in the context of a resource access request with access to the required resource(s). Only the requesting user could assume that role and they would be destroyed afterwards.

@marcoandredinis That was my first thought as well. But we are actually using the roles to define which resources you are allowed to search for, which logins you should have, how long the cert should last, etc. So they should be manually created in order to configure this, otherwise we have to add some new way to configure those options. Also we didn't want to add some new kind of transient auto-generated role, it would break assumptions, auditing, and automation that works with roles.

**Comment by marcoandredinis** [comment]
Got it, it makes sense
Thank you :pray:

**Pull request merged**